### ISSUE TITLE
Improve typings by providing `mutation`, `query`, `subscription` instead of `graphql`

### ISSUE THREAD

**Issue by RIP21** [dotansimha/graphql-code-generator#964]
Currently, HOCs for `apollo-react` generated using `graphql` which is quite obvious, but it actually has some flaws described in this issue:
https://github.com/apollographql/react-apollo/issues/2568

Long story short, both `Mutation` and `Query` function arguments are merged using `|` so both `mutation` and `data` end up in `props` arguments object.

To provide better typing you can import `query`, `mutation`, `subscription` from `react-apollo` and use them instead of `graphql`
E.g
```typescript
import { mutation } from 'react-apollo/mutation-hoc' // import directly, since not exported from root.

export function CreateMessageHOC<TProps, TChildProps = any>(
  operationOptions:
    | ReactApollo.OperationOption<
        TProps,
        CreateMessageMutation,
        CreateMessageVariables,
        CreateMessageProps<TChildProps>
      >
    | undefined,
) {
  return mutation< // mutation here instead of graphql
    TProps,
    CreateMessageMutation,
    CreateMessageVariables,
    CreateMessageProps<TChildProps>
  >(CreateMessageDocument, operationOptions)
```
This way typings are a little more correct.

Although this is hopefully will be addressed from `react-apollo` side as well if this will be merged
https://github.com/apollographql/react-apollo/pull/2573
It will change signature to `withMutation` etc. And will be exported from the root, so `ReactApollo.withMutation` will be possible.
 
You can fix it right away now, or you can postpone it until PR above is not merged and released.

**Comment by dotansimha** [comment]
@ardatan What do you think?

**Comment by ardatan** [comment]
I used `graphql` to avoid unused imports, because tsc can complain according to the project's tsconfig.
If we handle to eliminate unused imports etc, it is way better to use these hocs directly instead of struggling with generics.

**Comment by dotansimha** [comment]
@ardatan I think you can use the codegen core to expose `hasQueries`, `hasMutations` and `hasSubscriptions` here: https://github.com/dotansimha/graphql-code-generator/blob/master/packages/graphql-codegen-core/src/operations/transform-document.ts#L110-L111 and then it will make it easier to know which imports do you actually need.

**Comment by ardatan** [comment]
Thank you! I will take a look

**Comment by RIP21** [comment]
@dotansimha @ardatan https://github.com/apollographql/react-apollo/pull/2573
PR mentioned above has been merged. So `with*` thing will be real very soon.

**Comment by dotansimha** [comment]
Fix available in 1.0.0 👍
